url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+			    for (Integer edge : graph.get(no))
+				if ( !visited.contains(edge) )
+				    toVisit.add(edge);
+			}
+			
+			// is the number of visited nodes, greater than the 'currently' largest component?
+			if ( visited.size() > maximumSize ) { // it is! we've found a greater component then...
+			
+			    // update the maximum size
+			    maximumSize = visited.size();
+			    
+			    // update our greatest component
+			    greatestComponent = visited;
+			} else
+			    continue;
+		    }
+		    
+		    /*
+		     * we've found the maximum size connected component -- 'greatestComponent'
+		     */
+		    
+		    // let's choose the row
+		    for (Row row : rows.values()) {
+			// is it a node in the graph?
+			if ( row.edges != null ) { // it is
+			
+			    // get the nodes connected through this edge
+			    Integer[] edge = row.edges.toArray(new Integer[2]);
+			    int node1 = edge[0];
+			    int node2 = edge[1];
+			    
+			    // is this row an edge in the maximum size component?
+			    if ( greatestComponent.contains(node1) && greatestComponent.contains(node2) ) {
+				rLinha = row.id;
+				break;
+			    } else
+				continue;
+			} else
+			    continue;
+		    }
+		    
+		    chosenRowsCounter++;
+		} else { // no rows with 2 ones
+		    chosenRowsCounter++;
+		}
+	    }
+	    
+	    /*
+	     * a row has been chosen! -- 'rLinha'
+	     */
+	    
+	    // get the chosen row
+	    Row chosenRow = rows.get(rLinha);
+	    
+	    /*
+	     * "After the row is chosen in this step, the first row of A that intersects V is exchanged with the chosen row so that the chosen row is the first
+	     * row that intersects V."
+	     */
+	    
+	    // if the chosen row is not 'i' already
+	    if ( rLinha != i ) {
+		// swap i with rLinha in A
+		byte[] auxRow = A[i];
+		A[i] = A[rLinha];
+		A[rLinha] = auxRow;
+		
+		// swap i with rLinha in X
+		auxRow = X[i];
+		X[i] = X[rLinha];
+		X[rLinha] = auxRow;
+		
+		// decoding process - swap i with rLinha in d
+		int auxIndex = d[i];
+		d[i] = d[rLinha];
+		d[rLinha] = auxIndex;
+	    }
+	    
+	    /*
+	     * "The columns of A among those that intersect V are reordered so that one of the r nonzeros in the chosen row appears in the first column of V and
+	     * so that the remaining r-1 nonzeros appear in the last columns of V."
+	     */
+	    
+	    // if there are non-zeros
+	    if ( chosenRow.degree > 0 ) {
+		// stack of non-zeros in the chosen row
+		Stack<Integer> nonZeros = new Stack();
+		
+		// search the chosen row for the positions of the non-zeros
+		for (int nZ = 0, col = i; nZ < chosenRow.nonZeros; col++) // TODO the positions of the non-zeros could be stored as a Row attribute
+		{ // this would spare wasting time in this for (little optimization)
+		    if ( A[i][col] == 0 ) // a zero
+			continue;
+		    else { // a non-zero
+			nZ++;
+			
+			// add this non-zero's position to the stack
+			nonZeros.push(col);
+		    }
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1542:1552, 2425:2433.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.